### PR TITLE
Allow filtering on newly added columns

### LINKED ISSUE #217: Incorrect filtering on newly added columns
### Feature Request / Improvement

Reported on Slack: https://apache-iceberg.slack.com/archives/C029EE6HQ5D/p1702607510827899 Thanks for reporting this 👍 

Able to reproduce locally:
image

### PR DESCRIPTION
**Pull request by Fokko**
Resolves #217

### CODE REVIEW

File: pyiceberg/expressions/visitors.py [hunk 1]
```diff
@@ -906,7 +906,16 @@ def visit_bound_predicate(self, predicate: BoundPredicate[L]) -> BooleanExpressi
...
 def translate_column_names(expr: BooleanExpression, file_schema: Schema, case_sensitive: bool) -> BooleanExpression:
-    return visit(expr, _ColumnNameTranslator(file_schema, case_sensitive))
+    try:
+        return visit(expr, _ColumnNameTranslator(file_schema, case_sensitive))
+    except ValueError as e:
+        if "Not found in file schema" in str(e):
+            if isinstance(expr, BoundIsNull):
+                return AlwaysTrue()
+            else:
+                return AlwaysFalse()
```
Comment (amogh-jahagirdar) [inline pyiceberg/expressions/visitors.py:916]: Curious, why not do this in the visitor itself (_ColumnNameTranslator#visit_bound_predicate)? Is it something we would want decoupled from it?

File: tests/test_integration.py [hunk 2]
```diff
@@ -373,6 +379,15 @@ def test_scan_branch(test_positional_mor_deletes: Table) -> None:
     assert arrow_table["number"].to_pylist() == [1, 2, 3, 4, 6, 7, 8, 9, 10, 11, 12]
 
 
+@pytest.mark.integration
+def test_filter_on_new_column(test_table_add_column: Table) -> None:
+    arrow_table = test_table_add_column.scan(row_filter="b == '2'").to_arrow()
+    assert arrow_table["b"].to_pylist() == ['2']
```
Comment (amogh-jahagirdar) [inline tests/test_integration.py:385]: As a sanity check should we add a test which validates that a row_filter='b is NOT NULL' returns the same?

**Review by amogh-jahagirdar — APPROVED** [review]
Some questions, but nothing that I think is blocking, thanks @Fokko !

File: tests/test_integration.py (hunk 2 above)
Comment (Fokko) [reply tests/test_integration.py:385]: That's a good one, added 👍

File: pyiceberg/expressions/visitors.py (hunk 1 above)
Comment (Fokko) [reply pyiceberg/expressions/visitors.py]: I actually like that more, great suggestion 👍
